src/Stratis.Features.FederatedPeg/TargetChain/FederatedPegBroadcaster.cs

```diff
@@ -26,31 +25,19 @@ public class FederatedPegBroadcaster : IFederatedPegBroadcaster
         /// <inheritdoc />
         public async Task BroadcastAsync(Payload payload)
         {
-            // TODO: Optimize how federation peers are identified.
+            IEnumerable<INetworkPeer> connectedPeers = this.connectionManager.ConnectedPeers
+                .Where(peer => (peer?.IsConnected ?? false) && this.federatedPegSettings.FederationNodeIpEndPoints.Contains(peer.PeerEndPoint));
 
-            List<INetworkPeer> peers = this.connectionManager.ConnectedPeers.ToList();
-
-            var ipAddressComparer = new IPAddressComparer();
-
-            // TODO: Can do the send to each peer in parallel.
-
-            foreach (INetworkPeer peer in peers)
```

Comment:
@quantumagi Is this part where we were constantly looping over the node's ConnectedPeers list?

So is the change here that instead of checking inside the `foreach` we are now getting a list beforehand?

Couldn't an easier change have been to use the existing `Match` extension method here?

```
if (this.federatedPegSettings.FederationNodeIpEndPoints.Any(e => e.MatchIpOnly(peer.PeerEndPoint)))
```

Is the idea that we want to just broadcast to the federation nodes? Then why can't we just loop over the IPs in `this.federatedPegSettings.FederationNodeIpEndPoints` , check if each is connected and if so send the message?

Reply:
Looping serially makes the entire broadcast take potentially quite a while. If some of the peers are connected but temporarily unreachable the entire loop could take minutes. This part doesn't really have anything to do with the dictionary retrieval, it's just a nice to have.

Reply:
But yes, the idea is that we just want to broadcast to the federation nodes.

However, the ability to quickly identify whether an arbitrary peer is a federation member is broadly useful, so I think it's better doing it this way (via dictionary).

Reply:
Changing the broadcasting to a send the messages in parallel is a good change 👍 

However, all the other changes feels a bit overkill. I mean, we just want to loop over the federation nodes and if they are connected, send a message to them. 

What are we actually gaining by introducing a hashset? Its not like we are having to iterate over 10000's of IPs. Hence my suggestion of just to loop over the federation IPs, check if each one is connected (by calling FindByIp) and if so, send the message?

Reply:
> However, all the other changes feels a bit overkill. I mean, we just want to loop over the federation nodes and if they are connected, send a message to them.

If M is the number of nodes and N is the number of federation nodes then we want to improve on making M x N comparisons.

With the hash set we only do M comparisons, with hash set lookup only for the connected nodes.

Reply:
`FindByEndpoint` iterates over all members so I would rather have M operations than M x N operations.

Reply:
@quantumagi OK I hear ya 👍